lib/rubygems/sigstore/cert_chain.rb

```diff
@@ -0,0 +1,51 @@
+require "open-uri"
+require "rubygems/sigstore/cert_extensions"
+
+class Gem::Sigstore::CertChain
+  PATTERN = /-----BEGIN CERTIFICATE-----(?:.|\n)+?-----END CERTIFICATE-----/.freeze
+
+  def initialize(cert_pem)
+    @cert_pem = cert_pem
+  end
+
+  def certificates
+    @certificates ||= build_chain
+  end
+
+  def signing_cert
+    certificates.last
+  end
+
+  def root_cert
+    certificates.first
+  end
+
+  private
+
+  def build_chain
```

Comment:
Deserialize the current cert chain, and then iteratively:
- retrieve the leftmost certificate's parent via AIA coordinates
- deserialize the file into a Certificate instance
- prepend it into the chain

The looping stops once we see a cert having in AIA.  Checking whether it's actually a CA will be done during chain verification.